### PR TITLE
RSDK-2512 Add 'appimage' label to GitHub workflow

### PR DESCRIPTION
**Pull request by jeremyrhyde**
This PR adds the creation of PR-specific appimage (both x86_64 and arm64) upon request. The creation of these app images can be triggered by adding the `appimage` or `appimage-ignore-tests` labels. The `appimage` label will only begin creation if the test workflow has been completed successfully, while the `appimage-ignore-tests` label will produce them in parallel to the tests. 

In order to allow forked repos to generate app images, this PR also migrates us to a trusted workflow setup for PRs. This is highly desirable for a public repo as it will prevent malicious users from being able to initiate custom code, DDOS, or take up CPU. Instead, using the `pr-labeler.yml`, we first check if the github user is a viamrobotics person and only then allow them to run tests/appimage jobs.

The app images produced during this process will be stored in the google cloud along with the latest and stable images. Links to them will be proved by a github bot, as seen below. The links will take the form of:
```
https://packages.viam.com/apps/slam-servers/carto_grpc_server-pr-<PR NUMBER>-aarch64.AppImage
```

Elements of this workflow were taken:
-  RDK's [pullrequest-trusted.yml](https://github.com/viamrobotics/rdk/blob/main/.github/workflows/pullrequest-trusted.yml) - using `safe to test`, `appimage` and `appimage-ignore-tests` labels **[ALMSOT DIRECT COPY FROM RDK]**
- RDK's [pr-labeler.yml](https://github.com/viamrobotics/rdk/blob/main/.github/workflows/pullrequest-trusted.yml) - allow the use of GCP credentials for app image building from a fork as well as the automatic un-label/re-label to occur that allows the pullrequest-trusted.yml to be run. **[DIRECT COPY FROM RDK]**
-  RDK's [appimage.yml](https://github.com/viamrobotics/rdk/blob/main/.github/workflows/appimage.yml) - Checking out one repo versus another based on if the associated action is a pull, push or manual run
- RDK's [code-coverage-comment.yml](https://github.com/viamrobotics/rdk/blob/main/.github/workflows/code-coverage-comment.yml) - Producing gitbot messages
- RDK's [pullrequestclose.yml](https://github.com/viamrobotics/rdk/blob/main/.github/workflows/code-coverage-comment.yml) - Cleaning up appimages **[DIRECT COPY FROM RDK]**

Note 1: Up until now app images produced by our workflows have only be possible on code in main, this PR adds the ability in a PR to create an image using the code currently being pushed in order to enable both authors and reviewers to test their code more easily, assuming they are a viamrobotics memeber. 

Note 2: Gitbot comments with links will be removed and overwritten when a new run has been completed. The links will remain the same as they are based on the PR number but the time at which the comment was generated should allow users to know when they have the most up-to-date copy of an appimage for a particular PR

Note 3: Current code changes the `appimage.yml` from the one in main to one in this PR. This prevents testing from being done on this PR as it is currently watching for PRs targeting this branch. This will be reverted back to main before merge, allowing testing to occur.

**Testing:**
Code was altered in cartographer and tests ran showing the code had been updated. Additonally, tests were run when pushing withoutl abels, with an `appimage` label and with an `appimage-ignore-tests` label to confirm proper action (i.e. no appimage produced, appimage produced only after tests have run, appimage produced while test was still ongoing). finally, tests were done on both PRs from `viamrobotics/viam-cartographer` and a fork repo to ensure developers could trigger these.

This PR (https://github.com/viamrobotics/viam-cartographer/pull/51) Is available to test out the process.

The main.yml code was also tested here: https://github.com/viamrobotics/viam-cartographer/actions/runs/4745971047/jobs/8429023203

JIRA Ticket: https://viam.atlassian.net/browse/RSDK-2512

### CODE REVIEW

**Comment by nicksanford** [comment]
> _ No description provided. _

Ticket?

**Comment by nicksanford** [comment]
Also could you please link to how you figured out this workflow?

File: viam-cartographer_internal_test.go
Comment (nicksanford) [inline viam-cartographer_internal_test.go:143]: why is this changing?

File: .github/workflows/pullrequest.yml
Comment (jeremyrhyde) [reply .github/workflows/pullrequest.yml:19]: Yes I will once I can confirm a successful updated app image has been produced

File: .github/workflows/appimage.yml
Comment (jeremyrhyde) [inline .github/workflows/appimage.yml:109]: [note] Can add AppImage links in one comment but since the x86_64 takes longer there is the possibility that it can be clicked before it has been truly updated (or exists)

File: viam-cartographer_internal_test.go
Comment (jeremyrhyde) [reply viam-cartographer_internal_test.go]: This was a temp change for testing

File: .github/workflows/pr-labeler.yml
Comment (jeremyrhyde) [inline .github/workflows/pr-labeler.yml:1]: [note] RDK link: https://github.com/viamrobotics/rdk/blob/main/.github/workflows/pr-labeler.yml

File: .github/workflows/pullrequestclose.yml
Comment (jeremyrhyde) [inline .github/workflows/pullrequestclose.yml:1]: [note] https://github.com/viamrobotics/rdk/blob/main/.github/workflows/pullrequestclose.yml

File: .github/workflows/pullrequest-trusted.yml
Comment (jeremyrhyde) [inline .github/workflows/pullrequest-trusted.yml:1]: [note] RDK link https://github.com/viamrobotics/rdk/blob/main/.github/workflows/pullrequest-trusted.yml

Comment (JohnN193) [reply .github/workflows/pullrequest-trusted.yml:1]: Just to be clear, all instances of `qaworkflow` also have to be changed to main before merging right?

File: .github/workflows/pullrequestclose.yml
Comment (JohnN193) [reply .github/workflows/pullrequestclose.yml:1]: just to be clear, all instances of `qaworkflow` will have to be changed to main for this file, right?

File: .github/workflows/pr-labeler.yml
Comment (JohnN193) [reply .github/workflows/pr-labeler.yml:1]: before merging `qaworkflow` needs to be changed for this file as well right?

**Review by JohnN193 — APPROVED** [review]
LGTM, the only question I have is if you can test this with main.yml to ensure the main branch update still works as intended(manually run it on this branch, with `main.yml` pointing to `qaworkflow`).

Also it does not look like you cannot manually run the new github actions, but since these are copy/paste from rdk I think that's okay

File: .github/workflows/pr-labeler.yml
Comment (jeremyrhyde) [reply .github/workflows/pr-labeler.yml:1]: Yep, all references to qaworkflow will be removed before merge

**Comment by nicksanford** [comment]
> 

@Otterverse can you confirm that this is using pr-trusted.yml correctly? I want to confirm we are following best practices given that this change will now run PRs against viam owned infra.

File: .github/workflows/appimage.yml
Comment (nicksanford) [inline .github/workflows/appimage.yml:30]: When will `github.event_name == 'workflow_dispatch'` evaluate to true?

File: .github/workflows/appimage.yml
Comment (nicksanford) [inline .github/workflows/appimage.yml:40]: 1. When will `github.event_name == 'pull_request_target'` evaluate to true?

2. What is this ref doing?

File: .github/workflows/appimage.yml
Comment (nicksanford) [inline .github/workflows/appimage.yml:46]: Why do we need to add this step? Is this fixing a bug?

File: .github/workflows/appimage.yml
Comment (nicksanford) [inline .github/workflows/appimage.yml:88]: Why are we changing the order of authorizing GCP uploads & building images?

File: .github/workflows/appimage.yml
Comment (nicksanford) [inline .github/workflows/appimage.yml:130]: Is this mutually exclusive with https://github.com/viamrobotics/viam-cartographer/pull/46/files#diff-00f0606e18b2618c3cc1667ca7c2b703b537af690ca71eba1330633587dcb1eeR94?

File: .github/workflows/main.yml
Comment (nicksanford) [inline .github/workflows/main.yml:21]: Please remove after merge

File: .github/workflows/pr-labeler.yml
Comment (nicksanford) [inline .github/workflows/pr-labeler.yml:5]: Please remove after merge

File: .github/workflows/pullrequest-trusted.yml
Comment (nicksanford) [inline .github/workflows/pullrequest-trusted.yml:10]: Please remove before merge

File: .github/workflows/pullrequest-trusted.yml
Comment (nicksanford) [inline .github/workflows/pullrequest-trusted.yml:20]: Please remove before merge

File: .github/workflows/pullrequest-trusted.yml
Comment (nicksanford) [inline .github/workflows/pullrequest-trusted.yml:29]: Please remove before merge

File: .github/workflows/pullrequest-trusted.yml
Comment (nicksanford) [inline .github/workflows/pullrequest-trusted.yml:37]: Please remove before merge

File: .github/workflows/pullrequestclose.yml
Comment (nicksanford) [inline .github/workflows/pullrequestclose.yml:9]: Please remove before merge

File: .github/workflows/pr-labeler.yml
Comment (nicksanford) [reply .github/workflows/pr-labeler.yml:1]: Mind please posting the diff?

File: .github/workflows/pullrequest-trusted.yml
Comment (nicksanford) [reply .github/workflows/pullrequest-trusted.yml:1]: mind please posting the diff?

File: .github/workflows/pullrequestclose.yml
Comment (nicksanford) [reply .github/workflows/pullrequestclose.yml:1]: mind please posting the diff?

File: .github/workflows/pullrequest-trusted.yml
Comment (nicksanford) [inline .github/workflows/pullrequest-trusted.yml:19]: Shouldn't this be 
```suggestion
    if: (github.event.label.name == 'safe to test' && github.event.label.name == 'appimage')
```

**Comment by nicksanford** [comment]
1. What workflow steps now execute on viam owned infrastructure? 
2. Where are we prescribing that?

File: .github/workflows/appimage.yml
Comment (Otterverse) [reply .github/workflows/appimage.yml:30]: workflow_dispatch is the trigger when this workflow is called from another. E.g. when another workflow "uses" it.

File: .github/workflows/appimage.yml
Comment (Otterverse) [reply .github/workflows/appimage.yml:40]: pull_request_target is when this branch (normally "main") is the target of a PR. E.g. when someone wants to merge to main.

The ref is saying "check out the PR's code" (which may be untrusted) instead of the default of checking out the code in the context of the workflow (e.g. "main")

File: .github/workflows/appimage.yml
Comment (Otterverse) [reply .github/workflows/appimage.yml:46]: Best practice, but _technically_ should always be a clean checkout under a github runner.

File: .github/workflows/pullrequest-trusted.yml
Comment (Otterverse) [reply .github/workflows/pullrequest-trusted.yml:19]: No, because there are plenty of cases where you don't need/want to build an appimage, but you DO want to test.

File: .github/workflows/appimage.yml
Comment (Otterverse) [reply .github/workflows/appimage.yml:109]: You should be able to solve this pretty easily. Just have the "add links" bit be it's own job, and depend on the full appimage flow. Then it'll only run when both parts of the matrix are complete.

Or you can have another trigger for completion like we do in RDK (see the code coverage workflow, which also posts the appimage links.)

File: .github/workflows/appimage.yml
Comment (Otterverse) [inline .github/workflows/appimage.yml:86]: I know not technically your change, but there's no reason to use sudo inside the bash command string. Go tests shouldn't need to be root.

File: .github/workflows/main.yml
Comment (Otterverse) [inline .github/workflows/main.yml:20]: I think you should have a test job here first, and have appimage depend on it. You don't want to build new releases if they don't pass tests after merging.

**Review by Otterverse — CHANGES_REQUESTED** [review]
Looks pretty good, but I do think you should add testing to the main workflow and fix up the appimage link commenter to post things as one comment.

**Comment by JohnN193** [comment]
> Looks pretty good, but I do think you should add testing to the main workflow and fix up the appimage link commenter to post things as one comment.

We initially skipped having testing on main because testing would take an additional ~30 minutes to complete, and were flaky. Since that is no longer the case it def makes sense to add them back in

**Review by jeremyrhyde — COMMENTED** [review]
@Otterverse I'm going over your comments now James, but have two questions for you while I do so:
1. Where/when do the GCP credentials get set?
2. How does the current workflow, have here or in rdk, prevent the third party users from creating a PR and adding/removing labels, triggering the test and/or workflow events?

File: .github/workflows/appimage.yml
Comment (jeremyrhyde) [reply .github/workflows/appimage.yml:110]: Yes, these are mutually exclusive as 'workflow_dispatch' is a manual trigger from github.com, 'push' is the github event of a PR being merged and 'pull_request_target' is when a PR has been open and interacted with in one of several defined ways (labeled, opened, reopened, synchronized...). All events are mutually exclusive and if not, the concurrency group at the top of the pullrequest-trusted in concert with the !cancelled() check will prevent multiple runs from occurring simultaneously on the same PR.

File: .github/workflows/pullrequest-trusted.yml
Comment (jeremyrhyde) [reply .github/workflows/pullrequest-trusted.yml:19]: What about the instance in which a PR is labeled appimage but not safe to test? Should this be able to trigger test since they are required by the appimage job?

File: .github/workflows/appimage.yml
Comment (jeremyrhyde) [reply .github/workflows/appimage.yml]: slam integration tests need sudo access: https://github.com/viamrobotics/viam-cartographer/actions/runs/4756567063/jobs/8452292353?pr=51

Comment (jeremyrhyde) [reply .github/workflows/appimage.yml:86]: Note: motion-benchmarks.yml does this as well in RDK

**Comment by Otterverse** [comment]
> @Otterverse I'm going over your comments now James, but have two questions for you while I do so:
> 
> 1. Where/when do the GCP credentials get set?

There's a SECRET stored at the organization level. It contains the credentials.

> 2. How does the current workflow, have here or in rdk, prevent the third party users from creating a PR and adding/removing labels, triggering the test and/or workflow events?

Only users with the "Triage" permission can add/remove labels. External users don't have that on github.

File: .github/workflows/appimage.yml
Comment (Otterverse) [reply .github/workflows/appimage.yml:86]: I'm in the process of removing those cases in RDK, as they should not be doing that. One of the reasons we've got the "sudo -Hu testbot bash -lc ..." pattern everywhere is it forces thing to run as a normal user, and make thing break if they try to access stuff outside of where we expect and intend. If sudo is required here, then something else is wrong, and setup needs to be done differently. I expect the RDK code being checked out isn't being chown'd to the testbot user.

File: .github/workflows/appimage-comment.yml
Comment (Otterverse) [inline .github/workflows/appimage-comment.yml:16]: Where are you getting the env from here? I think you deleted a couple of steps that were required. See the full set of steps in https://github.com/viamrobotics/rdk/blob/main/.github/workflows/code-coverage-comment.yml

File: .github/workflows/appimage.yml
Comment (Otterverse) [inline .github/workflows/appimage.yml:45]: I'd add a leading line here
`chown -R testbot:testbot .` to change the ownership to the testbot user and prevent potential issues.

File: .github/workflows/main.yml
Comment (Otterverse) [reply .github/workflows/main.yml:24]: Where? It's still not showing a "needs" line for me. Maybe you forgot to push?

File: .github/workflows/appimage.yml
Comment (Otterverse) [reply .github/workflows/appimage.yml:86]: Yeah, chown the rdk code to testboth first, before switching to the testbot user for the actual test run.

File: .github/workflows/pullrequest-trusted.yml
Comment (Otterverse) [reply .github/workflows/pullrequest-trusted.yml:19]: If you demand something by adding a label, you should get it and not have to add TWO things to get one outcome. You made your choice. :-)

(Also, we do it this way in RDK, so consistency is good.)

**Review by Otterverse — CHANGES_REQUESTED** [review]
Down to two things, both mentioned inline. sudo/chown stuff and the missing steps in the commenter code.

File: .github/workflows/appimage.yml
Comment (jeremyrhyde) [reply .github/workflows/appimage.yml:45]: Why `chown -R testbot:testbot .` instead of `chown -R testbot .`?

**Review by nicksanford — APPROVED** [review]

File: .github/workflows/appimage-comment.yml
Comment (Otterverse) [reply .github/workflows/appimage-comment.yml:19]: This still needs the download of the pr.env artifact first. See https://github.com/viamrobotics/rdk/blob/main/.github/workflows/code-coverage-comment.yml#L13-L18 

It's a poorly named step, but the key thing is that it's getting "pr.env" (a file) from the original workflow, and that contains the APPIMAGE and PR_NUMBER settings you're using here.

Comment (Otterverse) [reply .github/workflows/appimage-comment.yml:19]: Okay, I see what's missing in understanding possibly. I didn't realize your test workflow didn't have the markers for this already.

See this: https://github.com/viamrobotics/rdk/blob/main/.github/workflows/test.yml#L85-L106

It's the other side of the comment system. It's what's setting the APPIMAGE and PR_NUMBER variables, storing them in pr.env, and then uploading pr.env in the first place. You'll need that part too. Sorry I didn't realize you were missing the other half here, or I would've explained better/sooner.

**Review by Otterverse — CHANGES_REQUESTED** [review]
Still some missing pieces for the comment stuff. I think hopefully I explained better now that I realized you were missing the early pieces as well. Poke me on slack if it's still not clear.

File: .github/workflows/test.yml
Comment (Otterverse) [inline .github/workflows/test.yml:116]:
```suggestion
    - name: Upload PR variables
```

File: .github/workflows/appimage-comment.yml
Comment (Otterverse) [inline .github/workflows/appimage-comment.yml:34]:
```suggestion
            <http://packages.viam.com/apps/slam-servers/carto_grpc_server-pr-${{ env.PR_NUMBER }}-aarch64.AppImage>
            <http://packages.viam.com/apps/slam-servers/carto_grpc_server-pr-${{ env.PR_NUMBER }}-x86_64.AppImage>
```

**Review by Otterverse — APPROVED** [review]
LGTM! Now that you're an expert on workflows, I'm gonna make you handle it the next  time they need a refactor in RDK! :-P (j/k)

**Pull request merged**
